### DISCUSSION TITLE
Invokable class as Eloquent event listener. How to type?

### DISCUSSION THREAD

**Discussion by rudiedirkx** [laravel/framework#53572] [Ideas]
I use event listeners on some models to log property changes:

```php
class User extends Model {
	static protected function booted() : void {
		self::updated(new ModelPropertiesObserver('user_changed', 'user_id', ['status', 'email', 'remote_id'], ['password', 'tfa_secret']));
	}
}
```

and that works, but it's not technically correct, because `HasEvents::updated()` expects `\Illuminate\Events\QueuedClosure|\Closure|string|array` for callback, not an invokable class.

Is there a better way to do that?, or could Laravel accept invokable classes? It could add `|callable` to the `@param`, but that seems too wide. Maybe a new Laravel specific Eloquent event interface?

```php
interface InvokableEventHandler {
	public function __invoke(Model $object) : void;
}
```

**Comment by henzeb** [comment]
Why would callable be too wide ?

**Reply by rudiedirkx** [reply]
I don't know, that's up to Taylor I guess. Maybe it's good enough, but it is very wide. The `string` that's hinted is not a callable string, but a class name. Maybe `callable` would be confusing?

What about the invokable class? Is that an acceptable event handler like this? Maybe there's an obvious better more-Laravel way that I missed.

**Reply by shaedrich** [reply]
That would raise the question if it should be `class-string` rather than `string` 🤔

**Comment by rudiedirkx** [comment]
Yaaaay https://github.com/laravel/framework/pull/53642
